File: src/main/scala/rps/GameService.scala
package rps

import java.util.UUID

import rps.model._
import zio.IO
import scala.util.Random
import model._
import Move._
import Result._

trait GameService {
  def playMove(userMove: Move): IO[Throwable, UUID]
  def getResult(): IO[Throwable, Option[Play]]
}

class GameServiceImpl(repository: GameRepository) extends GameService {
  override def playMove(userMove: Move): IO[Throwable, UUID] = {
    val computerMove = generateComputerMove()
    val result = (userMove, computerMove) match {
      case (Rock, Scissors) | (Paper, Rock) | (Scissors, Paper) => Win
      case (x, y) if x == y                                     => Draw
      case _                                                    => Lose
    }    
    val play = Play(userMove, computerMove, result)

    repository.save(play)
  }

  private def generateComputerMove(): Move =
    Random.shuffle(List(Rock, Paper, Scissors)).head

  override def getResult(): IO[Throwable, Option[Play]] =
    repository.read()
}

File: rps-ts/src/services/gameService.ts
import { Move, Error, Result, PlayResponse } from "../models";
import { TaskEither, right } from "fp-ts/lib/TaskEither";
import { GameRepository } from "../repository";

export interface GameService {
  play: (move: Move) => TaskEither<Error, Result>;
  lastPlay: () => TaskEither<unknown, PlayResponse>;
}

const randomMove = (): Move => {
  const random = Math.floor(Math.random() * 3) + 1;
  switch (random) {
    case 1:
      return "Rock";
    case 2:
      return "Paper";
    default:
      return "Scissor";
  }
};

export const createGameService = (repo: GameRepository): GameService => ({
  play: (move: Move): TaskEither<Error, Result> => {
    const computerMove = randomMove();
    if (move === computerMove) return right("Draw");
    switch (move) {
      case "Paper":
        return right(computerMove === "Rock" ? "Win" : "Lose");
      case "Rock":
        return right(computerMove === "Scissor" ? "Win" : "Lose");
      case "Scissor":
        return right(computerMove === "Paper" ? "Win" : "Lose");
    }
  },
  lastPlay: () => repo.getLast,
});
